deps/chakrashim/src/v8debug.cc

```diff
@@ -21,71 +21,715 @@
+      // already exist.
+      {
+        Context::Scope context_scope(debugContextRef);
+        jsrt::ContextShim* contextShim = jsrt::ContextShim::GetCurrent();
+        JsValueRef ensureDebug = contextShim->GetensureDebugFunction();
+        JsValueRef unused;
+        if (jsrt::CallFunction(ensureDebug, *global, &unused) != JsNoError) {
+          return Local<Context>();
+        }
+      }
+    }
+
+    return static_cast<Context*>(isoShim->debugContext->GetContextRef());
+  }
 
-    Local<Context> currentContext = Context::GetCurrent();
-    if (!currentContext.IsEmpty()) {
-      // Turn on debug mode on current Context (script engine), which was
-      // created before start debugging and not in debug mode.
-      JsStartDebugging();
+  void CALLBACK JsDiagDebugEventHandler(
+    JsDiagDebugEvent debugEvent,
+    JsValueRef eventData,
+    void* callbackState) {
+    if (Debug::handler == nullptr) {
+      return;
     }
+
+    Debug::messageQueue.debugEventProcessCount++;
+    bool isMsgToProcess = !Debug::messageQueue.IsEmpty();
+    bool shouldContinue = true;
+    do {
+      if (eventData != nullptr) {
+        Debug::messageQueue.ProcessDebugEvent(debugEvent, eventData);
+        eventData = nullptr;
+      }
+
+      // Process all messages received from debugger
+      if (Debug::messageQueue.debugEventProcessCount <= 1) {
+        isMsgToProcess = !Debug::messageQueue.IsEmpty();
+        while (!Debug::messageQueue.isProcessingDebuggerMsg && isMsgToProcess) {
+          Debug::messageQueue.ProcessDebuggerMessage();
+          isMsgToProcess = !Debug::messageQueue.IsEmpty();
+        }
+      }
+
+      shouldContinue = Debug::messageQueue.ShouldContinue(debugEvent);
+      if (!shouldContinue) {
+        Debug::messageQueue.WaitForMessage();
+      }
+
+      isMsgToProcess = !Debug::messageQueue.IsEmpty();
+    } while (isMsgToProcess &&
+      !Debug::messageQueue.isProcessingDebuggerMsg && !shouldContinue);
+
+    Debug::messageQueue.debugEventProcessCount--;
+  }
+
+  void Debug::StartDebugging(JsRuntimeHandle runtime) {
+    JsErrorCode errorCode = JsDiagStartDebugging(
+      runtime, JsDiagDebugEventHandler, nullptr);
+    CHAKRA_VERIFY_NOERROR(errorCode);
+  }
+
+  void Debug::SendCommand(Isolate* isolate,
+    const uint16_t* command, int length,
+    ClientData* client_data) {
+    CHAKRA_ASSERT(client_data == NULL);
+
+    // Save command in a queue and when JsDiagDebugEventHandler
+    // is called we need to process the queue
+    messageQueue.SaveMessage(command, length);
+
+    // Request async break from engine so that we can process
+    // commands on JsDiagDebugEventAsyncBreak
+    JsErrorCode errorCode = JsDiagRequestAsyncBreak(
+      jsrt::IsolateShim::FromIsolate(isolate)->GetRuntimeHandle());
+    CHAKRA_VERIFY_NOERROR(errorCode);
   }
 
-error:
-#else
-  hr = E_FAIL;  // ChakraCore does not support JsStartDebugging
+  static JsValueRef CALLBACK Log(
```

Comment:
>Log [](start = 29, length = 3)

If i understand correctly is this for debugging the debugger code and will never called from user scenario?